clients/venice-admin-tool/src/main/java/com/linkedin/venice/ZkCopier.java

```diff
@@ -0,0 +1,196 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.commons.io.FileUtils;
+
+
+public class ZkCopier {
+  private static final Set<String> VENICE_ZK_PATHS = new HashSet<>(
+      Arrays.asList(ADMIN_TOPIC_METADATA, EXECUTION_IDS, PARENT_OFFLINE_PUSHES, ROUTERS, STORE_GRAVEYARD, STORES));
+
+  /**
+   * Extract Venice-specific paths from an input text file to an output text file.
+   * @param inputPath absolute path to the ZK snapshot input text file, which contains one path per line
+   * @param outputPath absolute path to output text file with extracted Venice-specific paths, which contains one path per line
+   * @param clusterNames set of cluster names
+   * @param basePath base path for ZK
+   */
+  public static void extractVenicePaths(
+      String inputPath,
+      String outputPath,
+      Set<String> clusterNames,
+      String basePath) {
+    try {
+      // read ZK snapshot input file and store all paths in a list
+      File inputFile = new File(inputPath);
+      List<String> zkPaths = FileUtils.readLines(inputFile);
+
+      // write Venice-specific paths to output file
+      ArrayList<String> venicePaths = getVenicePaths(new ArrayList<>(zkPaths), clusterNames, basePath, VENICE_ZK_PATHS);
+      File outputFile = new File(outputPath);
+      FileUtils.writeLines(outputFile, venicePaths);
+    } catch (IOException e) {
+      throw new VeniceException(e.getMessage());
+    }
+  }
+
+  /**
+   * Get Venice-specific paths filtered by base path, cluster names, and required paths from a list of ZK paths.
+   */
+  public static ArrayList<String> getVenicePaths(
+      ArrayList<String> zkPaths,
+      Set<String> clusterNames,
+      String basePath,
+      Set<String> requiredPaths) {
+    TreeNode requiredPathsTree = buildRequiredPathsTree(clusterNames, basePath, requiredPaths);
+    TreeNode zkPathsTree = listToTree(zkPaths, basePath);
+    getVenicePathsHelper(zkPathsTree, requiredPathsTree);
+    return treeToList(zkPathsTree);
+  }
+
+  private static void getVenicePathsHelper(TreeNode zkPathsTree, TreeNode requiredPathsTree) {
+    if (requiredPathsTree.getChildren().isEmpty() || zkPathsTree.getChildren().isEmpty()) {
+      return;
+    }
+    ArrayList<String> TreeNodesToBeDeleted = new ArrayList<>();
+    for (TreeNode zkChild: zkPathsTree.getChildren().values()) {
+      if (!requiredPathsTree.contains(zkChild.getVal())) {
+        TreeNodesToBeDeleted.add(zkChild.getVal());
+      }
+    }
+    for (String str: TreeNodesToBeDeleted) {
+      zkPathsTree.deleteChild(str);
+    }
+    for (TreeNode requiredChild: requiredPathsTree.getChildren().values()) {
+      if (zkPathsTree.contains(requiredChild.getVal())) {
+        getVenicePathsHelper(zkPathsTree.getChildren().get(requiredChild.getVal()), requiredChild);
+      }
+    }
```

Comment:
This seems a bit convoluted, and does a two-pass check. We should be able to simplify this by doing a one-pass through the `children`. 
1. iterate through `zkPathsTree.entrySet()`
2. Check if `key` is present in `requiredPathsTree.getChildren()` map
    1. If it is present, recurse `getVenicePathsHelper` on it
    2. If it is not present, `zkPathsTree.deleteChild(key)`

We don't really need to maintain a `TreeNodesToBeDeleted`. However, if we end up using `TreeNodesToBeDeleted`, it should start with lower case.

Reply:
Thank you for the hint! I encountered an `ConcurrentModificationException` which I thought was because we were removing an element from the map while also iterating through the map. I resolved this issue using `Iterator`.